README.md
![Codeship Status for vpgambit/collabot](https://codeship.com/projects/87838d20-936b-0132-849a-5668b62cd35b/status?branch=dev)

Collabot v. 0.9.5
=================
A fun slack bot to incentive collaboration between developers.


Running it
==========

- Set the mongoDB variables located at config.js in your ~/.bashrc file.
- Make sure you specify appropriate values to environment variables required. The mechanism to specify environment variables depends on the OS or shell you're using, so please read the related documentation if you're not sure about how to do it. See below a sample set of variables for reference.
- Type `npm install` and then `npm test` to finally issue `npm start`. 
- Then do a GET to the URL specified in the APPSTART environment variable - that will log the bot into Slack. 


Other tasks
===========

Collabot works over Slack API integration. It's configured by default on one channel, but it can be set to join multiple channels:

* Any user can add the bot to any channel 
* The admins can set the channels the bot is in via the bot integration panel

Sample Environment Variables (*nix shell example)
=================================================
``` bash
# Gambit!
export ENVIRONMENT="Gambit dev's Lap"
export APPURL="http://localhost:3000/"
export MDBURL="mongodb://localhost:27017/collabot"
export CHANNEL="gambit-test"
export BOTNAME="alpha-bot"
export BOT_TOKEN="xoxb-4181420889-GK8qfKC5tnr2RerqcGfsg55P"
export BOT_REGIONAL='{"medellin": {"groups": ["gambit-disposable-01"], "manager": "slizarazo", "channels" : {"training" : "gambit-disposable-02"}}}'
```

Changelog.md
What's New in v0.9.5
====================
- Fixed: commands that post to channels go crazy when issued from DM

What's New in v0.9.4
====================
- when a new training is created, a small "teaser" message is sent to default channel, while details are posted to training channel (city-specific)

What's New in v0.9.3
====================
- registration threshold is now 1 h
- a reminder for a forgetful presenter is also created (notification to present te session!)


What's New in v0.9.0
====================
- Scheduled jobs for notifications are persistent (bot dies overnight)
- As an HR user I can see the results of the rating process for each session
- FIXED: When creating a training, you cannot specify a date in the future
- Disabled node-monkey and enabled node-debug


What's New in v0.8.4
====================
- FIXED: When giving points, the messages should appear in the channel where the action occurs, not in #general

What's New in v0.8.3
====================
- Added "show attendants" in the help command list, adjusted the behaviors and messages

What's New in v0.8.2
====================
- Fixed: registration threshold is being applied inversely

What's New in v0.8.0
====================
- Send reminders to register, to attend and to rate.

* Registration Reminder
 * Sent once to all members of the community `$edserv.reminders.register.hours` (18) hours before the event.
* Event reminder
 * Sent once to all of the event's registered members `edserv.reminders.attend.hours` (8) hours before the event takes place.
* Rating reminder
 * Sent once to all members who have not yet rated after `$edserv.reminders.rate.hours` (1) hours 
 
What's New in v0.7.5
====================
- Fixed the conversation for create training session (it was awkwardly waiting for an extra line from the user before starting)

What's New in v0.7.4
====================
- When a presenter is entering a training's time, only the format `HH:MM` should be accepted

What's New in v0.7.3
====================
- Do not display past training sessions when "show me upcoming sessions" is invoked

What's New in v0.7.2
====================
- Add the command to publish (to a configurable channel only) the link to COPASO  gdoc template

What's New in v0.7.1
====================
- A bot is not a player, so points assigned to bot shouldn't be valid

What's New in v. 0.7.0
======================
- As an user I can rate sessions that I have attended

What's New in v. 0.5.5
======================
- When giving only 1 point, it won't expect the word "points"
- You don't need to name "collabot" in all lowercase to get bot's attention

What's New in v. 0.5.4
======================
- support addressing users with @ when giving points
- when giving 1 point, you are not required to say "points"

What's New in v. 0.5.3
======================
- Added support for specifying a reason when giving points

What's New in v. 0.5.2
======================
- Bot can now join private groups (only those it's been invited to)

What's New in v. 0.5.1
======================
- You cannot assign points to yourself anymore. Hey! That's just fair!

What's New in v. 0.5.0
======================
- Environment variables further independence
- As an HR user I can see a list of attendants to a session

What's new in v. 0.4.0
======================
- Environment variables isolation to ease release process (Env variable for "environment")
- An user can create (register) a new training (BnL) session with all the information associated to it
- An user can register (enroll) to an existing, upcoming BnL session
- Stabilization of dependencies versions and adde dinfo about the project in package.json

What's new in v. 0.3.0
======================
- A separate set of infrastructure for PROD was created (stability enhanced)
- An HR user can authorize another user to be a presenter

What's New in v. 0.2.0
======================
- Filters for 'top' operation are available: Bot top now takes two optional arguments: a period of time (day, week, month, year) and a channel. Those values are used to filter top scores data.
- web tier now finds the app endpoints using config file or environment properties (appUrls)
- minor bug fixes

What's New in v. 0.1.0
======================
- Collabot will support ony one instance in backend (guid)
- Web interface uses now a switch component instead of two buttons
- commands now require the bot name as specified in config file
- staged conversation with the bot is now supported (greeting)


config.js
var IRCConnector = require("./classes/IRCConnector.class");
var SlackConnector = require("./classes/SlackConnector.class");
var JSONConnector = require("./classes/persistence/JSONConnector.class");
var MongoConnector = require("./classes/persistence/MongoConnector.class");
var regional = JSON.parse(process.env.BOT_REGIONAL);

// Slack API config karmabot for tests
module.exports = {
	version: "0.9.5",
	environment: process.env.ENVIRONMENT,
	// environment: "rick's lap",
	channel: process.env.CHANNEL,
	botName: process.env.BOTNAME,
	// token: 'xoxb-3558335570-MXw76pJPGbNLLyLi1KYWK8qr', // Good ol' Collabot
	token: process.env.BOT_TOKEN, //'xoxb-3749826050-mKrPHbMYqRDa0mb47lhf3sfr', // Rick's tal-bot 
	autoReconnect: true,
	autoMark: true,
	connector: SlackConnector,
	persistence: MongoConnector,
	// dbURL: 'mongodb://'+process.env.MDBUSERNAME+':'+process.env.MDBPWD+'@'+process.env.MDBHOST+':'+process.env.MDBPORT+'/'+process.env.MDBDB,
	dbURL: process.env.MDBURL,// 'mongodb://localhost:27017/collabot',
	appUrls: {
		start: process.env.APPURL+'start', 
			// 'http://salty-inlet-8617.herokuapp.com/start',
			// 'http://localhost:3000/start',
		stop: process.env.APPURL+'stop'
			// 'http://salty-inlet-8617.herokuapp.com/stop'
			// 'http://localhost:3000/stop'
	},
	maxCollabPoints : 10,
	edserv: {
		thresholds: {
			enrollment: 1
		},
		reminders: {
			register: {hours: 18},
			attend: {hours: 8},
			rate: {hours: 1, afterEvent: true, retry: {times: 3, intervalHours: 24}}
		},
		manager: 'slizarazo',
		regional: regional
	},
	copaso: {
		group: 'vpc-copaso', 
		template: 'https://docs.google.com/spreadsheets/d/15BmXmYVE0Yc7704VgV_N8vGypr4vFWoOiX-rfHN0MDg'
	}
};

// Slack API config prod
// module.exports = {
// 	environment: "platanoMachine",
// 	channel: "karma",
// 	botName: "karmabot",
// 	token: 'xoxb-3534020363-hXwUqB5krA8fMZbxR7VmHsxB',
// 	autoReconnect: true,
// 	autoMark: true,
// 	connector: SlackConnector,
// 	persistence: MongoConnector,
//	dbURL: 'mongodb://localhost:27017/slashbot'
// };

// Sample Slack-thru-IRC config
// module.exports = {
// 	environment: "platanoMachine",
// 	channel: "foundrystories",
// 	server: "ffoundry.irc.slack.com",
// 	botName: "obibot",
// 	password: "ffoundry.MUr0IDVbPois5VgPtbYY",
// 	connector: SlackConnector,
// 	token: 'xoxb-3463721915-mXLDStOphmDcva9aBNBuYCcT',
// 	autoReconnect: true,
// 	autoMark: true,
// 	persistence: JSONConnector
// };

// Sample plain IRC config

// module.exports = {
// 	environment: "slashieMachine",
// 	channel: "#slashbot",
// 	server: "verne.freenode.net",
// 	botName: "obibot",
// 	connector: IRCConnector,
// 	persistence: MongoConnector,
// 	dbURL: 'mongodb://localhost:27017/slashbot'
// };